File: python/809. Expressive Words.py
class Solution(object):
    def expressiveWords(self, S, words):
        """
        :type S: str
        :type words: List[str]
        :rtype: int
        """
        def check(W):
            ns, nw = len(S), len(W)
            i = j = 0
            l = 1
            while i < nw and j < ns:
                if W[i] != S[j]:
                    return False
                i += 1
                j += 1
                stepw = steps = 0
                while i < nw and W[i-1]==W[i]:
                    stepw += 1
                    i += 1
                while j < ns and S[j-1]==S[j]:
                    steps += 1
                    j += 1
                if stepw > steps:
                    return False
                if steps > stepw and steps < 2:
                    return False
            return i==nw and j==ns

        return len(filter(check, words))
        
if __name__ == '__main__':
    S = "heeellooo"
    words = ["hello", "hl", "helo"]
    su = Solution()
    print su.expressiveWords(S, words)

File: go/src/leetcode/809/809. Expressive Words.go
/*
*
@author ZhengHao Lou
Date    2022/11/25
*/
package main

import "fmt"

/*
*
https://leetcode.cn/problems/expressive-words/
*/
func expressiveWords(s string, words []string) (tot int) {
	var target [][]int

	for i := range s {
		j := int(s[i] - 'a')
		if len(target) == 0 || target[len(target)-1][0] != j {
			target = append(target, []int{j, 1})
		} else {
			target[len(target)-1][1]++
		}
	}
OUT:
	for _, word := range words {
		var k int
		var i int
		for i < len(word) {
			j := i + 1
			for j < len(word) && word[j] == word[j-1] {
				j++
			}
			l := j - i
			if k >= len(target) || target[k][0] != int(word[i]-'a') ||
				target[k][1] < l || (target[k][1] != l && target[k][1] < 3) {
				//fmt.Println(target[k], l)
				i = j
				continue OUT
			}
			i = j
			k++
		}
		if k == len(target) {
			tot++
		}
	}
	fmt.Println(tot)
	return
}

func main() {
	expressiveWords("heeellooo", []string{"hello", "hi", "helo"})
}
